### PR TITLE
The Pattern-Matching Bug: introduce a temporality heuristic to de-pessimize certain programs

### PR DESCRIPTION
**Pull request by gasche**
(This PR sits on top of #13152 and #13153; for now it is best reviewed by looking only at the latest commit.)

#13152 pessimizes Total matches in transitively mutable position. It gives two examples of code that now generates worse code:

```ocaml
(* case (A) *)
let f : bool option ref -> ... = function
| { contents = None } -> ...
| { contents = Some true } -> ...
| _ when guard () -> ...
| { contents = Some false } (* here *) -> ...

type _ gadt = Int : int t | Bool : bool t

(* case (B) *)
let g : int gadt ref -> ... = function
| { contents = Int } (* here *) -> ...
```

In the case A, the generated code first checks something about the argument root.0, then it checks a guard, then it goes back to the argument root.0. (We say that there are two "splits", resulting in three submatrices, that are checked in sequence.) When it goes back to root.0, its value could have changed in the meantime, so assuming that root.0.0 could only be `false` is wrong, and we really needed to make the code worse.

In the case B, the generated code reads the mutable field root.0, then immediately checks all possible cases on this constructor. In this case there cannot have been any mutation "in the meantime", this is the first time ever that we are checking this value. So generating a `Match_failure` case here was not necessary, and we are generating worse code than we could.

In general it is hard to know for sure "this is the first time we do something on this value". Tracking this precisely might be done with more context information, but it would be a tricky change with quickly diminishing returns. There is however a case where it is very easy to see that we are in this situation: where we have not generated any "split", any logic of the form "check this about the first value, and if that fails check that".

The present PR introduces a new piece of static context information used by the pattern-matching compiler to track this criterion. In code:

```ocaml
type temporality =
  | First
  | Following
(** The [temporality] information tracks information about the
    placement of the current submatrix within the
    whole pattern-matching.

    - [First]: this is the first submatrix on this position seen by values
      that flow into the submatrix.
    - [Following]: there was a split, some other submatrix was tried first
      and failed, and the control jumped to the current submatrix.

    This information is used in {!compute_arg_partial}.
*)
```

When deciding whether a switch on a given position should be pessimized, we now use a refined criterion: we are in a transitively mutable position *and* the current temporality is `Following`, we have already accessed this position before.

In particular, the example `g` above now produces good code again, as well as any reasonable usage of GADTs in mutable positions that I can think of.

It is, of course, possible to build examples where the heuristic is too coarse-grained and which will generate worse code than they could. For example:

```ocaml
type _ gadt = Int : int t | Bool : bool t

let test : int gadt ref -> unit = function
  | _ when Random.bool () -> ()
  | { contents = Int } -> ()
```

### CODE REVIEW

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml]: Very minor nit: I would suggest avoiding interrupting the arity of `c_rec`, given it is always applied to three arguments.

File: lambda/matching.ml
Comment (ncik-roberts) [inline lambda/matching.ml]:
```suggestion
   The criterion that we use is based on two contextual informations:
```

**Review by ncik-roberts — APPROVED** [review]
The approach looks correct. In a survey of a large codebase, I found only one instance where this PR de-pessimizes a program -- I would consider arguing against this change if it significantly complicated the matching.ml code, but given that the change is clear and well-structured, I'm in favor of merging it.

**Comment by gasche** [comment]
One reason I like this heuristic is that it helps me reason about "structural matches" that do not contain splits -- all possible cases are explored in full. For example:

```ocaml
function
| None -> 0
| Some {contents = A} -> 1
| Some {contents = B _} -> 2
| Some {contents = C _} -> 3
```

(assuming the type-checker finds this example exhaustive: `A`, `B of ...`, `C` are the only valid constructors at the specific type involved -- it could be a GADT with other, incompatible constructors.)

For this fragment of pattern-matching programs, the type-checker exhaustiveness analysis is always correct, and the temporality heuristic guarantees that we generate good code.

**Comment by gasche** [comment]
Rebasing this on top of #13338 (transitively) made the implementation quite a bit less invasive, because I can add the temporality as an extra field to the `partiality` record instead of passing it along as before.

File: testsuite/tests/match-side-effects/check_partial.ml
Comment (gasche) [inline testsuite/tests/match-side-effects/check_partial.ml]: I did not understand why this test case changed -- there was a Match_failure case before and there is still one after, so what did the temporality change in the code? This is much easier to see in the `-drawlambda` output for this test:

```diff
(setglobal Test!
  (let
    (lazy_needs_partial/274 =
       (function param/276 : int
         (catch
           (let (*match*/279 =a (field_imm 0 param/276))
             (catch
               (let
                 (*match*/280 =a (field_imm 1 param/276)
                  *match*/281 =o (field_mut 0 *match*/280))
-                 (if (isint *match*/281) (if *match*/281 (exit 2) 0)
-                   (exit 1)))
+                 (switch* *match*/281 case int 0: 0
+                                      case int 1: (exit 2)))
               with (2)
               (let
                 (*match*/288 =
                    (let
                      (lzarg/282 = *match*/279
                       tag/283 =a (caml_obj_tag lzarg/282))
                      (if (== tag/283 250) (field_mut 0 lzarg/282)
                        (if (|| (== tag/283 246) (== tag/283 244))
                          (apply (field_imm 1 (global CamlinternalLazy!))
                            (opaque lzarg/282))
                          lzarg/282)))
                  *match*/289 =a (field_imm 1 param/276)
                  *match*/290 =o (field_mut 0 *match*/289))
                 (if (isint *match*/290) (if *match*/290 12 (exit 1))
                   (exit 1)))))
         with (1)
           (raise
             (makeblock 0 (global Match_failure/20!) [0: "test.ml" 6 49])))))
    (makeblock 0 lazy_needs_partial/274)))
```

This is about the way that the check `(_, {contents = True})` is compiled in the first clause. The GADT type declaration is as follows:

```ocaml
type _ t =
  | Int : int -> int t
  | True : bool t
  | False : bool t
```

Before the compiled would do first an `isint` test to rule out `Int`, and then an `if` test to distinguish `True` from `False`. But with this commit, the compiler now trusts the totality information that tells it that the `Int` case is impossible, and generates a simpler check to differentiate `True` from `False`. (There is a similar check down below, after the `lazy ()` pattern, and that ones is pessimized as it should.)

**Review by lpw25 — APPROVED** [review]
Approving on the basis of @ncik-roberts review

**Comment by gasche** [comment]
Thanks! The present PR sits on top of #13341, so I should wait for that PR to be merged before merging the present one.

**Comment by gasche** [comment]
This PR can be merged now that #13341 has been approved and merged. I am planning to do this after/if the CI passes.

**Pull request merged**
